source/develop/release-management/features/sla/soft-host-to-vm-affinity-support.html.md

```diff
@@ -0,0 +1,120 @@
+,enforcing -> vm_enforcing ,vds_positive,vds_enforcing ,vms_affinity_enabled (true/false  - default false)**
+
+>**NOTE** : The additional “vms_affinity_enabled”  flag is needed for affinity group to be able to express that VMs in an affinity group have no relationship to each other.
+
+>**NOTE** :  for each affinity group there must be at least one entry with a vm_id in affinity_group_members table (even only for a vm to host affinity).
+
+
+### Scheduler
+*   Modify VmAffinity filter and weight policy units to exclude affinity groups with vms_affinity_enabled = false;
+*   Add new filter policy unit **VmToHostAffinityFilterPolicyUnit** which:
+Enables Affinity Groups hard enforcement for VMs to hosts; VMs in group are required to run either on one of the hosts in group (positive) or on independent hosts which are excluded from the hosts in group (negative).                              
+*   Add new weight policy unit **VmToHostAffinityWeightPolicyUnit** which:
+Enables Affinity Groups soft enforcement for VMs to hosts; VMs in group are most likely to run either on one of the hosts in group (positive) or on independent hosts which are excluded from the hosts in group (negative)
+*   Add new Load balancer - vms To Hosts Affinity Balancer  (Future development to be done) 
+
+#### Affinity Rules Enforcement Manager
+Enhance or rewrite AREM to properly enforce violated affinity rules
+TBD
+
+#### REST API
+The current solution will be enhanced to support the additional hosts list and the new attributes:
+vm_positive,vm_negative,vds_positive,vds_negative,vms_affinity_enabled
+
+Need to garuntee that the old API still works.
+
+See example (new additions are in red) :
+
+GET /api/clusters/00000002-0002-0002-0002-000000000222/affinitygroups
+
+```xml
+
+<affinity_groups>
+  <affinity_group href="/ovirt-engine/api/clusters/00000002-0002-0002-0002-000000000222/affinitygroups/31ef70c1-e636-45a6-9492-aa4fad753e6f" id="31ef70c1-e636-45a6-9492-aa4fad753e6f">
+    <name>Test_aff_group</name>
+    <link href="/ovirt-engine/api/clusters/00000002-0002-0002-0002-000000000222/affinitygroups/31ef70c1-e636-45a6-9492-aa4fad753e6f/vms" rel="vms"/>
+    <link href="/ovirt-engine/api/clusters/00000002-0002-0002-0002-000000000222/affinitygroups/31ef70c1-e636-45a6-9492-aa4fad753e6f/vms" rel="hosts"/>
+    <cluster href="/ovirt-engine/api/clusters/00000002-0002-0002-0002-000000000222" id="00000002-0002-0002-0002-000000000222"/>     
+    <vm_positive>true</vm_positive>
+    <vm_enforcing>true</vm_enforcing>
+    <vds_positive>true</vds_positive>
+    <vds_enforcing>true</vds_enforcing>
+    <vms_affinity_enabled>true</vms_affinity_enabled>
+  </affinity_group>
+</affinity_groups>
+```
+>**NOTE** : vm_enforcing and vm_positive tags were originally enforcing and positive.
+> in order to perserve old functionallity either : 
+> 1) the old name will be kept and represent vms. 
+> 2) vm_enforcing and vm_positive will be added and positive/enforcing tags will be deprecated
```

Comment:
I think that option 2) is better.

As you are introducing a set of cofigurations that are identical for virtual machines and hosts, I'd suggest to introduce a new `AffinityGroupConfiguration` type (you can probably find a better naem) grouping them. Then you can have two attributes of that type, one for virtual machines and another for hosts:

```xml
<vms_configuration>
  <positive>true|false</positive>
  <enforcing>true|false</enforcing>
  <enabled>true|false</enforcing>
</vms_configuration>
<hosts_configuration>
  <positive>true|false</positive>
  <enforcing>true|false</enforcing>
  <enabled>true|false</enforcing>
</hosts_configuration>
```

The existing `positive` and `enforcing` attributes need to be preserved, and marked as deprecated. They should have exactly the same meaning that they had before.